### ISSUE TITLE
coding standard should explain the motivation for our choice of shebang line

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3768]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3768|
|Type|enhancement|
|Created|2009-04-10 16:12:20Z|
|Branch|https://github.com/twisted/twisted/tree/shebang-std-3768-2|

In [#3750](https://github.com/twisted/twisted/issues/3750), the coding standard has recently been changed to require `/usr/bin/env python` rather than `/usr/bin/python`.

The coding standard should document the motivations for this choice, so that somebody doesn't switch it back again in the future.  It should also lay out a few of the pros and cons of using `/usr/bin/python` vs. `/usr/bin/env` in the shebang line, and explain that distutils will rewrite it upon installation so this policy only covers the source files in version control.

<details><summary>Searchable metadata</summary>

```
trac-id__3768 3768
type__enhancement enhancement
reporter__glyph glyph
priority__highest highest
milestone__ 
branch__branches_shebang_std_3768_2 branches/shebang-std-3768-2
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation__easy documentation, easy
time__1239379940000000 1239379940000000
changetime__1247443769000000 1247443769000000
version__None None
owner__ 
cc__exarkun cc__thijs cc__glyph cc__mwh
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Since I think this should have been done as _part_ of [#3750](https://github.com/twisted/twisted/issues/3750), I'd really like to see it get taken care of right away, not fall through the cracks for six years.

thijs, any interest in tackling this?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Agreed.  Poke poke, thijs.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra set status to `assigned`|
|-|-|

yep, will start on it in the next days

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

(In [26682]) Branching to 'shebang-std-3768'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Note http://twistedmatrix.com/trac/ticket/3750#comment:7

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Thijs, I don't think we're probably not going to block the release on this, but nevertheless I would _really_ appreciate it if this were complete in time for 9.0, which should be happening soon.  Is the branch anywhere close to review?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

(In [27099]) Branching to 'shebang-std-3768-2'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

(In [27100]) Add motivation for our choice of shebang line, refs [#3768](https://github.com/twisted/twisted/issues/3768)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_10"></a>|@mwhudson set owner to @thijstriemstra|
|-|-|

Cool to see all these doc improvements.

I think the "and explain that distutils will rewrite it upon installation so this policy only covers the source files in version control." part of the ticket is quite valuable and doesn't seem to be present, unless I'm missing something.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [27127]) Add note about distutils, refs [#3768](https://github.com/twisted/twisted/issues/3768)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra removed owner|
|-|-|

Replying to [mwh](#note_10):
> Cool to see all these doc improvements.
>
> I think the "and explain that distutils will rewrite it upon installation so this policy only covers the source files in version control." part of the ticket is quite valuable and doesn't seem to be present, unless I'm missing something.

I added this line, although I'm not sure what this means so I copied it verbatim. If you have a more complete description let me know and I'll add it, unless it's fine with the last change.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_13"></a>|@mwhudson set owner to @thijstriemstra|
|-|-|

I think this is a definite improvement, please land.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [27128]) Merge shebang-std-3768-2: documenting the motivations for Twisted's choice of
shebang line.

Author: thijs
Reviewer: mwh
Fixes: [#3768](https://github.com/twisted/twisted/issues/3768)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
